contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+import "../GenArt721CoreV2_PBAB.sol";
+
+import "@openzeppelin-4.5/contracts/utils/cryptography/ECDSA.sol";
+import "./ISignVerifierRegistry.sol";
+
+pragma solidity 0.8.9;
+
+/**
+ * @title Powered by Art Blocks ERC-721 core contract.
+ * Support for IYK pull mechanism to transfer tokens based on verified signatures
+ * @notice Due to the pull mechanism (claimNFT) being the only supported transfer function, these NFTs will NOT be tradeable on secondary marketplaces
+ * @author Art Blocks Inc.
+ */
+contract GenArt721CoreV2_IYK is GenArt721CoreV2_PBAB {
+    using ECDSA for bytes32;
+
+    // Address of registry that resolves the signVerifier
+    ISignVerifierRegistry public signVerifierRegistry;
+    bytes32 public signVerifierId;
+
+    // Nonce per user, used to prevent replay signature attacks
+    mapping(address => uint256) public claimNonces;
```

Comment:
> Seems a bit more straightforward and easier for the signVerifier to keep straight?

Could you elaborate on this comment? The `signVerifier` doesn't really need to keep track of any state on its own, it's always fetching from the contract before signing the claiming hash.

Also, keeping it per-user instead of per-token means that the following situation might occur:
1. Person A scans the item and gets a valid signature with token transfer nonce 1
2. Person B scans the item and gets a valid signature with token transfer nonce 1
3. Person A submits the signature to the contract
4. Person B can no longer submit their signature, even though they're the latest possessor of the item.

The above situation is avoided if signatures are submitted immediately but due to fluctuations in gas person A's txn may not go through until later and we don't want to prevent person B from also claiming the NFT.

Reply:
That is a good point - I was thinking of the following situation based on the current state of this PR:

1. Person A scans the item and gets a valid signature with Address A transfer nonce 1
2. Person B scans the item and gets a valid signature with Address B transfer nonce 1
3. Person B submits the signature to the contract and sees that they own the NFT
4. Person A submits the signature to the contract and now owns the token, while Person B is unaware that they no longer own the token

The remedy to both of our described situations is that person B can re-scan and re-claim their NFT. However, in this example, person B's ownership is custodial (they have to routinely check that Person A has not claimed their token), whereas in the token-transfer-nonce model, person B's ownership is non-custodial. Additionally, in this example, there could be ~infinite addresses between Person A and B (e.g. Person A.1, A.2, A.3, A.4, ...A.`n`), whereas in the token transfer nonce, the denial of service could only happen once.

All of this is likely mitigated by timestamp expiration of the signatures though, so definitely up to you regarding which model is preferred!

Reply:
resolving this 👍